augur/utils.py

```diff
@@ -12,12 +12,11 @@
 from pkg_resources import resource_stream
 from io import TextIOWrapper
 from .__version__ import __version__
-import packaging.version as packaging_version
-from .validate import validate, ValidateError, load_json_schema
 
 from augur.util_support.color_parser import ColorParser
 from augur.util_support.date_disambiguator import DateDisambiguator
 from augur.util_support.metadata_file import MetadataFile
+from augur.util_support.node_data_reader import NodeDataReader
```

Comment:
I'm in favor, but I'm not sure `NodeDataReader` as it appears in this branch is the best class to expose.

If we look at Augur from a domain-driven design perspective (as I understand DDD, anyway), the most important concept is the `NodeData` structure.  It is the data type that most of the Augur subcommands use for input and output.

It makes sense to me that the subcommands should be able to take and return these structures. But `NodeData` is kind of an internal name that requires understanding of how Augur works in order to make sense. A nicer public-facing term might be something like `DataSet` or `AugurData`—something that doesn't require the user to know about Nodes.

So a session might look something like this:

```python
from augur import DataSet

data = DataSet("/path/to/data.fasta")

filtered_data = data.filter(sequences_per_group=20)
filtered_data.write("/path/to/intermediate_data.fasta")

aligned_tree = filtered_data.align(reference="/path/to/reference.fasta") \
                            .tree()
```

The methods of `DataSet` would hide any additional complexity from the user—the API should be no more complicated than writing a Snakefile. Or even simpler, since you have the option of chaining calls without writing intermediate files. (Is this a use case that makes sense? As a non-user myself, I don't understand the value of Augur in a context without Snakemake.)

Since `DataSet` would (could) store all of the intermediate values, it could enforce required operation order:

```python
data = DataSet("/path/to/data.fasta")

data.tree()
# AugurSubcommandOrderError: tree() requires data to be aligned first
```

---

Does Augur even have enough users who rely on the Python API to justify supporting it as a first-class feature?

Forgive me if you've already been through these discussions internally, but I see it as a product management question. What do you want Augur to be: a command-line tool, a Python library, both, more? It can be any of those, but each one costs resources. It might be a reasonable decision to say that Augur is a CLI tool only, and that direct usage of any of the Python modules is unsupported.

Does the API allow meaningful usage that the CLI does not? Is Augur even usable through the API alone? I'm guessing it's inconvenient at best—many subcommand modules make assumptions about things like `args`, which are not things an API user would typically be required to care about.

The subcommands are already designed to use files as intermediate data stores. This is exactly what I would expect for a CLI tool, but unconventional for an API. See the hypothetical session above. It's going to require a substantial amount of work to provide that kind of "good" API experience. Is it worth the investment?

Reply:
Good questions @elebow -- I'll answer first but @huddlej will also have good opinions.

Augur has had a few iterations over the years, and initially was better thought of as a collection of functions which were combined into one-script-per-pathogen. I hesitate to use the term API here as we didn't really document it and it changed frequently. This had a number of problems: new pathogens required a new (and rather complicated) script, debugging was hard as the script run-time was long and no intermediate files were saved, and the code was hard to adapt to different use-cases as it was primarily build for influenza which has it's own specific needs.

We began scheming augur as it currently stands in early 2018, essentially aiming for the CLI based tool we now have. The big advantages here are the ability to quickly build new pipelines for pathogens (often with snakemake, but not necessarily so), to have the confidence to add / remove layers to existing builds, and to be able to combine it with other tools outside of augur (or custom scripts etc). 

So for me, it's primarily a CLI tool, however I know @huddlej also uses it in an API-capacity for his research.

Reply:
> What do you want Augur to be: a command-line tool, a Python library, both, more?

augur is primarily a command line tool, but we also use its Python interface as a small library of helper functions. Your comment here, @elebow, is very helpful in framing this scope though. We've discussed this topic a bit more internally recently and have decided to formally flag the Python functions we consider to be part of our public API and update our docs to state that any functions outside of that API are subject to change.

For now we'll hold off including these new helper classes as part of that public API, but we're open to introducing data-specific classes like a `NodeData` class when the use cases for them are well-justified.